Support for AWS STS for HCP CLI (#6496)
open-cluster-management/rhacm-docs: clusters/hosted_control_planes/create_hosted_aws.adoc

[#create-hosted-multi-zone-aws]
= Creating a hosted cluster in multiple zones on AWS

You can create a hosted cluster in multiple zones on AWS by using Security Token Service (STS) authentication. Complete the following steps:

. Create a JSON file named `trust-relationship.json` that contains the trust relationship configuration for your ARN role. See the following example:

+
[source,json]
----
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": <arn> <1>
            },
            "Action": "sts:AssumeRole"
        }
    ]
}
----

+
<1> Replace `<arn>` with the user name or the role that you used to generate the AWS session token.

> . Create the Amazon Resource Name (ARN) role. Replace `<name>` with the ARN role and `<file_name>` with the file name. Run the following command:

Reviewer: ```suggestion
. Create the IAM role. Replace `<name>` with any name. e.g. `hcp-cli-role` and `<file_name>` with the file name. Run the following command:
```

+
[source,bash]
----

> aws iam create-role --role-name <name> --assume-role-policy-document file://<file_name>.json

Reviewer: this command will have a long output. to capture the ARN value that we need to pass for `--role-arn` flag, we could do this:

```suggestion
ROLE_ARN=$(aws iam create-role --role-name <name> --assume-role-policy-document file://<file_name>.json --query "Role.Arn")
```

					"iam:DeleteInstanceProfile",
					"iam:TagInstanceProfile",
					"iam:GetRole",
					"iam:UpdateAssumeRolePolicy",
					"iam:GetInstanceProfile",
					"iam:TagRole",
					"iam:RemoveRoleFromInstanceProfile",
					"iam:CreateRole",
					"iam:DeleteRole",
					"iam:PutRolePolicy",
					"iam:AddRoleToInstanceProfile",
					"iam:CreateOpenIDConnectProvider",
					"iam:TagOpenIDConnectProvider",
					"iam:ListOpenIDConnectProviders",
					"iam:DeleteRolePolicy",
					"iam:UpdateRole",
					"iam:DeleteOpenIDConnectProvider",
					"iam:GetRolePolicy"
				],
				"Resource": "*"
			},
			{
				"Sid": "Route53",
				"Effect": "Allow",
				"Action": [
					"route53:ListHostedZonesByVPC",
					"route53:CreateHostedZone",
					"route53:ListHostedZones",
					"route53:ChangeResourceRecordSets",
					"route53:ListResourceRecordSets",
					"route53:DeleteHostedZone",
					"route53:AssociateVPCWithHostedZone",
					"route53:ListHostedZonesByName"
				],
				"Resource": "*"
			},
			{
				"Sid": "S3",
				"Effect": "Allow",
				"Action": [
					"s3:ListAllMyBuckets",
					"s3:ListBucket",
					"s3:DeleteObject",
					"s3:DeleteBucket"
				],
				"Resource": "*"
			}
		]

> ----

Reviewer: the policy needs to be attached to the role
```shell
aws iam put-role-policy \
    --role-name Test-Role \
    --policy-name ExamplePolicy \
    --policy-document file://policy.json
```

[#create-hosted-multi-zone-aws]
= Creating a hosted cluster in multiple zones on AWS

You can create a hosted cluster in multiple zones on AWS by using Security Token Service (STS) authentication. Complete the following steps:

. Create a JSON file named `trust-relationship.json` that contains the trust relationship configuration for your ARN role. See the following example:

+
[source,json]
----
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": <arn> <1>
            },
            "Action": "sts:AssumeRole"
        }
    ]
}
----

+

> <1> Replace `<arn>` with the user name or the role that you used to generate the AWS session token.

Reviewer: ```suggestion
<1> Replace `<arn>` with the ARN of the user or the role that you used to generate the AWS session token.
```

. Create the Identity and Access Management (IAM) role by running the following command:

+
[source,bash]
----
aws iam create-role \
--role-name <name> \ <1>
--assume-role-policy-document file://<file_name>.json \ <2>
--query "Role.Arn"
----

+

> <1> Replace `<name>` with the ARN role name, for example, `hcp-cli-role`.

Reviewer: ```suggestion
<1> Replace `<name>` with the role name, for example, `hcp-cli-role`.
```
the ARN is only an identifier generated by AWS after the resource (iam role) is created.

					"iam:DeleteInstanceProfile",
					"iam:TagInstanceProfile",
					"iam:GetRole",
					"iam:UpdateAssumeRolePolicy",
					"iam:GetInstanceProfile",
					"iam:TagRole",
					"iam:RemoveRoleFromInstanceProfile",
					"iam:CreateRole",
					"iam:DeleteRole",
					"iam:PutRolePolicy",
					"iam:AddRoleToInstanceProfile",
					"iam:CreateOpenIDConnectProvider",
					"iam:TagOpenIDConnectProvider",
					"iam:ListOpenIDConnectProviders",
					"iam:DeleteRolePolicy",
					"iam:UpdateRole",
					"iam:DeleteOpenIDConnectProvider",
					"iam:GetRolePolicy"
				],
				"Resource": "*"
			},
			{
				"Sid": "Route53",
				"Effect": "Allow",
				"Action": [
					"route53:ListHostedZonesByVPC",
					"route53:CreateHostedZone",
					"route53:ListHostedZones",
					"route53:ChangeResourceRecordSets",
					"route53:ListResourceRecordSets",
					"route53:DeleteHostedZone",
					"route53:AssociateVPCWithHostedZone",
					"route53:ListHostedZonesByName"
				],
				"Resource": "*"
			},
			{
				"Sid": "S3",
				"Effect": "Allow",
				"Action": [
					"s3:ListAllMyBuckets",
					"s3:ListBucket",
					"s3:DeleteObject",
					"s3:DeleteBucket"
				],
				"Resource": "*"
			}
		]
----

> . Attach the `policy.json` file to your ARN role by running the following command:

Reviewer: ```suggestion
. Attach the `policy.json` file to your role by running the following command:
```

+
[source,bash]
----
aws iam put-role-policy \

> --role-name <arn_role> \ <1>

Reviewer: ```suggestion
--role-name <role_name> \ <1>
```

--policy-name <policy_name> \ <2>
--policy-document file://policy.json <3>
----

+

> <1> Replace `<arn_role>` with the name of your ARN role.

Reviewer: ```suggestion
<1> Replace `<role_name>` with the name of your role.
```

<2> Replace `<policy_name>` with your policy name.

> <3> The `policy.json` file contains the ARN role policies.

Reviewer: ```suggestion
<3> The `policy.json` file contains the role permission policies.
```
